Change the default of conv.assert_shape flags to speed up compilation time

nouiz opened the pull request:
This happen in corner cases with scan from 20m to 1m.

@f0k How bad you think that we don't assert user shape by default? @abergeron point of view is that this is more for debugging purpose. We won't generate bad results. It would just put back the user shape as compiler hint as it was done in some of the ops (while others was enforcing it). We will be consistent between all ops.

see gh-5642

nouiz commented:
I should add, that my personal point of view is that as it cause big inconvenience, even in "corner" cases, we should not enable it by default.

Why does it cause such big slow down with scan? It is due in big part to the order of optimization. We push out nodes from the scan on the not optimized graph, that contain many duplicated nodes that aren't yet merged. We could pre-optimize the inner graph before the push out, but this is much more complicated and we won't do that so close to a release.

See also: https://github.com/Theano/Theano/pull/5635/commits/c52b93961e4143af356800f1ade0ca3efb408a00 and https://github.com/Theano/Theano/pull/5635/commits/5c2d16704ab31e5ba71e0f8a2db7427c4519a379

f0k commented:
> How bad you think that we don't assert user shape by default?

I don't think this is important. Most convolution implementations don't require knowing the shape in advance anyway. +1 for reducing compilation time (not only for scan, also for regular convnets).

IIRC, @gvtulder also added assertions for invalid shape combinations (e.g., shapes that would lead to negative output sizes). Are those still there? Or were they implemented as part of the convolution implementations, not as assertions?

f0k commented on theano/configdefaults.py:
```diff
@@ -130,7 +130,7 @@ def __str__(self):
     "If False, AbstractConv* ops won't add assert that verify that"
     " the user provided shapes are also the one at run time",
-    BoolParam(True),
+    BoolParam(False),
```
Maybe also update the documentation to something like "If True, AbstractConv* ops will verify that user-provided compile time shapes match the runtime shapes (debugging option, may slow down compilation)".

nouiz commented:
I updated the doc. I removed "compile time" as it isn't at compilation
time, but when building the graph. I don't think making the disctinction in
which phase is useful, so simpler seem better.

There is other assert for negative shapes. They are still there.

On Thu, Mar 2, 2017 at 9:50 AM Jan Schlüter <notifications@github.com>
wrote:

> *@f0k* commented on this pull request.
> ------------------------------
>
> In theano/configdefaults.py
> <https://github.com/Theano/Theano/pull/5643#discussion_r103940285>:
>
> > @@ -130,7 +130,7 @@ def __str__(self):
>      'conv.assert_shape',
>      "If False, AbstractConv* ops won't add assert that verify that"
>      " the user provided shapes are also the one at run time",
> -    BoolParam(True),
> +    BoolParam(False),
>
> Maybe also update the documentation to something like "If True,
> AbstractConv* ops will verify that user-provided compile time shapes match
> the runtime shapes (debugging option, may slow down compilation)".
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5643#pullrequestreview-24747837>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-1MEiFWbrFGVpEj4IP9UzhCz1iwHks5rhtckgaJpZM4MRDjJ>
> .
>

f0k commented:
> I removed "compile time" as it isn't at compilation time, but when building the graph.

Well yes, but what you give on building the graph is what's there at compile time, and what will end up in the generated code. So I usually just distinguish compile-time from runtime. Anyway, simpler is better!

abergeron commented on theano/configparser.py:
```diff
@@ -98,15 +98,17 @@ class change_flags(object):
+    def __init__(self, *args, **kwargs):
         confs = dict()
-        for k in kwargs:
+        args = dict(*args)
```
It might be better if we accepted a single dictionary argument in addition to the kwargs.

nouiz replied:
I'm not sure this is clear enough for me. Do you mean keeping *args as a parameter, then make sure it is only 1 element and it is a dict?

If so, this would be more restrictive then accepting anything that get converted to a dict. I don't have a strong preference. I just don't understand what it would bring to be more restrictive here except that the code would be more similar, but would make it harder for people to remember and probably look up more frequently what is accepted.

lamblin replied:
I find it confusing to have `*args` here, but to require it to only be 1 argument at most (otherwise the call to `dict` fails).
Why not `def __init__(self, d=None, **kwargs):` and then `if not d: d = {}; else d = dict(d)`?
I won't hold the release for that, though.

nouiz replied:
I think you are miss reading. Here there isn't any constrain on how many args there is. The only constrain is that dict accept it, so that we can convert it to a dict.

The line 108 with the assert of len(l)==1, is there to make sure the key only correspond to 1 Theano flag. In fact, if this is false, the problem isn't with the parameter args, but with Theano itself. We should never have 2 Theano flags with the same name.

abergeron commented on theano/configparser.py:
```diff
@@ -98,15 +98,17 @@ class change_flags(object):
     Useful during tests.
     """
-    def __init__(self, **kwargs):
+    def __init__(self, args=[], **kwargs):
```
Don't use a list as a default argument.  I know it doesn't really matter here, but it's a bad habit and a bad example.

lamblin replied:
Maybe an empty tuple?

lamblin commented on bin/theano-test:
```diff
@@ -1,3 +0,0 @@
-#!/usr/bin/env python
-import theano_test
-theano_test.main()
```
If you remove those files, you have to remove references to it from `setup.py` as well.

lamblin replied:
And from `MANIFEST.in`.

nouiz commented:
I removed the extra commit from this PR. It was already in another PR: https://github.com/Theano/Theano/pull/5651

I'll do the comment related to that in its own PR.

f0k commented:
> It would just put back the user shape as compiler hint as it was done in some of the ops

That's an important point, actually. For things like meta-optimization, we may need a way of specifying an example shape to optimize for, without asserting that this will be the only possible shape at runtime. +2 for not using the assertion by default!
